src/writer/stripe.rs

```diff
@@ -0,0 +1,175 @@
+            number_of_rows: Some(value.row_count as u64),
+            encrypt_stripe_id: None,
+            encrypted_local_keys: vec![],
+        }
+    }
+}
+
+/// Encode a stripe. Will encode columns into an in-memory buffer before flushing
+/// entire stripe to the underlying writer.
+pub struct StripeWriter<W> {
+    writer: W,
+    /// Flattened columns, in order of their column ID.
+    columns: Vec<Box<dyn ColumnStripeEncoder>>,
+    pub row_count: usize,
+}
+
+impl<W> EstimateMemory for StripeWriter<W> {
+    /// Used to estimate when stripe size is over threshold and should be flushed
+    /// to the writer and a new stripe started.
+    fn estimate_memory_size(&self) -> usize {
+        self.columns.iter().map(|c| c.estimate_memory_size()).sum()
+    }
+}
+
+impl<W: Write + Seek> StripeWriter<W> {
+    pub fn new(writer: W, schema: &SchemaRef) -> Self {
+        let columns = schema.fields().iter().map(create_encoder).collect();
+        Self {
+            writer,
+            columns,
+            row_count: 0,
+        }
+    }
+
+    /// Attempt to encode entire [`RecordBatch`]. Relies on caller slicing the batch
+    /// to required batch size.
+    pub fn encode_batch(&mut self, batch: &RecordBatch) -> Result<()> {
+        // TODO: consider how to handle nested types (including parent nullability)
+        for (array, encoder) in batch.columns().iter().zip(self.columns.iter_mut()) {
+            encoder.encode_array(array)?;
+        }
+        self.row_count += batch.num_rows();
+        Ok(())
+    }
+
+    /// Flush streams to the writer, and write the stripe footer to finish
+    /// the stripe. After this, the [`StripeWriter`] will be reset and ready
+    /// to write a fresh new stripe.
+    pub fn finish_stripe(&mut self) -> Result<StripeInformation> {
+        let start_offset = self.writer.stream_position().context(IoSnafu)?;
+
+        // Order of column_encodings needs to match final type vec order.
+        // (see arrow_writer::serialize_schema())
+        // Direct encoding to represent root struct
+        let mut column_encodings = vec![ColumnEncoding::Direct];
+        let child_column_encodings = self
+            .columns
+            .iter()
+            .map(|c| c.column_encoding())
+            .collect::<Vec<_>>();
+        column_encodings.extend(child_column_encodings);
+        let column_encodings = column_encodings.iter().map(From::from).collect();
+
+        // Root type won't have any streams
+        let mut written_streams = vec![];
+        for (index, c) in self.columns.iter_mut().enumerate() {
+            // Offset by 1 to account for root of 0
+            let column = index + 1;
+            let streams = c.finish();
+            // Flush the streams to the writer
+            for s in streams {
+                let (s_type, bytes) = s.into_parts();
+                let length = bytes.len();
+                self.writer.write_all(&bytes).context(IoSnafu)?;
```

Comment:
How about managing the content length ourselves to avoid calling `seek`?

Reply:
Need for Seek removed from writer :+1: